frontend/views/pages/GroupChat.vue

```diff
@@ -186,12 +200,25 @@ export default ({
         this.refreshTitle()
       })
       const { chatRoomID } = to.params
+      const { mhash } = to.query
       const prevChatRoomId = from.params.chatRoomID || ''
-      if (chatRoomID && chatRoomID !== prevChatRoomId) {
-        this.updateCurrentChatRoomID(chatRoomID)
-        // NOTE: No need to consider not-joined private chatroom because it's impossible
-        if (!this.isJoinedChatRoom(chatRoomID)) {
-          this.loadLatestState(chatRoomID)
+      if (chatRoomID) {
+        if (chatRoomID !== prevChatRoomId) {
+          if (!this.isJoinedChatRoom(chatRoomID)) {
+            if (this.hasPermissionToReadChatRoom(chatRoomID)) {
+              this.updateCurrentChatRoomID(chatRoomID)
+              this.loadLatestState(chatRoomID)
+            } else {
+              alert(L('Sorry, this message is from a private chatroom that you are not part of.'))
+              this.$router.go(-1)
+            }
+          } else {
+            this.updateCurrentChatRoomID(chatRoomID)
+          }
+        } else if (mhash) {
+          this.$refs.chatMain?.scrollToMessage(mhash).then(() => {
+            this.$router.replace({ query: {} })
+          })
```

Comment:
When user clicks the link to the message of the current chatroom, it is handled in here.